dali/kernels/imgproc/resample/separable_impl.h

```diff
@@ -0,0 +1,144 @@
+    ScratchpadEstimator se;
+
+    // Sample descriptions need to be delivered to the GPU - hence, the storage
+    se.add<SampleDesc>(AllocType::GPU, setup.sample_descs.size());
+
+    // CPU block2sample lookup may change in size and is large enough
+    // to mandate declaring it as a requirement for external allocator.
+    size_t num_blocks = setup.total_blocks.pass[0] + setup.total_blocks.pass[1];
+    se.add<SampleBlockInfo>(AllocType::GPU, num_blocks);
+    se.add<SampleBlockInfo>(AllocType::Host, num_blocks);
+
+    // Request memory for intermediate storage.
+    se.add<IntermediateElement>(AllocType::GPU, setup.intermediate_size);
+
+    req.scratch_sizes = se.sizes;
+    req.output_shapes = { setup.output_shape };
+    return req;
+  }
+
+  template <int which_pass, typename PassOutputElement, typename PassInputElement>
+  void RunPass(
+      const OutListGPU<PassOutputElement, 3> &out,
+      const InListGPU<PassInputElement, 3> &in,
+      const SampleDesc *descs_gpu,
+      const InTensorGPU<SampleBlockInfo, 1> &block2sample,
+      cudaStream_t stream) {
+    BatchedSeparableResample<which_pass>(
+        out.data, in.data, descs_gpu, in.num_samples(),
+        block2sample.data, block2sample.shape[0],
+        setup.block_size,
+        stream);
+  }
+
+  /// @remarks This function shall not allocate memory by ano other means
+  ///          than through `context.scratchpad`
+  virtual void
+  Run(KernelContext &context, const Output &out, const Input &in, const Params &params) {
+    cudaStream_t stream = context.gpu.stream;
+
+    SampleDesc *descs_gpu = context.scratchpad->Allocate<SampleDesc>(
+        AllocType::GPU, setup.sample_descs.size());
+
+    int total_blocks = setup.total_blocks.pass[0] + setup.total_blocks.pass[1];
+
+    OutTensorCPU<SampleBlockInfo, 1> sample_lookup_cpu = {
+      context.scratchpad->Allocate<SampleBlockInfo>(AllocType::Host, total_blocks),
+      { total_blocks }
+    };
+    OutTensorGPU<SampleBlockInfo, 1> sample_lookup_gpu = {
+      context.scratchpad->Allocate<SampleBlockInfo>(AllocType::GPU, total_blocks),
+      { total_blocks }
+    };
+    setup.InitializeSampleLookup(sample_lookup_cpu);
+    copy(sample_lookup_gpu, sample_lookup_cpu, stream);  // NOLINT (it thinks it's std::copy)
+
+    cudaMemcpyAsync(
+        descs_gpu,
+        setup.sample_descs.data(),
+        setup.sample_descs.size()*sizeof(SampleDesc),
+        cudaMemcpyHostToDevice,
+        stream);
+
+    InTensorGPU<SampleBlockInfo, 1> first_pass_lookup = make_tensor_gpu<1>(
+        sample_lookup_gpu.data,
+        { setup.total_blocks.pass[0] });
+
+    InTensorGPU<SampleBlockInfo, 1> second_pass_lookup = make_tensor_gpu<1>(
+        sample_lookup_gpu.data + setup.total_blocks.pass[0],
+        { setup.total_blocks.pass[1] });
+
+    intermediate.data = context.scratchpad->Allocate<IntermediateElement>(
+        AllocType::GPU, setup.intermediate_size);
+
+    RunPass<0, IntermediateElement, InputElement>(
```

Comment:
Maybe some names to easily distinguish passes, `0` and `1` are not that descriptive